modules/dists/BlockDist.chpl

```diff
@@ -1538,9 +1538,62 @@ where this.sparseLayoutType == unmanaged DefaultDist &&
   return true;
 }
 
+proc BlockArr.canDoOptimizedSwap(other) {
+  var domsMatch = true;
+  var domShapesMatch = true;
+
+  if this.dom != other.dom { // no need to check if this is true
+    if domsMatch {
+      for param i in 0..this.dom.rank-1 {
+        if this.dom.whole.dim(i) != other.dom.whole.dim(i) {
+          if this.dom.whole.dim(i).size != other.dom.whole.dim(i).size {
+            domShapesMatch = false;
+          }
```

Comment:
I'd probably be inclined to remove this `domShapesMatch=false` conditional and the subsequent "shapes match and they match their bounding boxes" conditional because it feels like a very specific special case that's unlikely to happen often in practice (i.e. I don't think typical code patterns result in this pattern).  So it feels like a bit of runtime overhead and generated code + needing to maintain / ensure correctness of this code without any obvious benefit.

A check I could imagine being more useful would be to see whether each locale involved in the swap owned a chunk of the same size, but that has greater execution time cost (so I don't think we should do it either, at least without a compelling use case).

Reply:
I don't feel strongly about it and can remove that check, but my thinking was that this and the subsequent check are executed only if corresponding domain dimensions don't match exactly. This implies that without these checks we'll run a `forall` to do the swap and the cost of these checks will be negligible compared to that, and that comparison only makes sense if these checks fail, too. Because otherwise, we'll get the benefit of pointer swapping.

It is common to have domains match their distributions' bounding boxes but you are suggesting that swaps between such arrays are not common, right?:

```chapel
var D1 = {1..10} dmapped Block({1..10});
var D2 = {0..9} dmapped Block({0..9});
var A1: [D1] int;
var A2: [D2] int;
A1 <=> A2;
```

In any case, left to my own devices, I'd keep the check for the reason above and that it adds coverage for a case that might come up. But I don't feel strongly about it as I see your points, too. I'll leave the final judgement call to you if you still think covering these cases doesn't worth the additional overhead on the unoptimized case.

Reply:
> but you are suggesting that swaps between such arrays are not common, right?:

Right.  To me the common cases are "we're all using the same domain or distribution" or "we're each using our own domain and/or distribution, but all using the same indexing scheme", but not "we've each got our own domain and distribution and indexing scheme" (while also wanting to swap arrays and expecting it to be fast).

I agree that the execution time checks are going to be negligible.  To me, the overhead is the additional code complexity in the swap operator that future developers would need to understand when modifying/updating this code (or, in the moment, that I need to feel confident in as a reviewer).

Given that mix, I'd still be inclined to remove it, as I'd prefer to avoid carrying around too much code optimizing for cases that don't have a use case that we'd consider compelling / likely / something that anyone should naturally expect to be fast.  Though if you disagree with that characterization (i.e., you think it is more likely than I expect or something that a user should naturally expect to be fast), I'm definitely open to counterarguments.

Reply:
All sounds reasonable to me and I buy the argument that maintenance overhead is important here. I don't have any concrete counterargument at the moment, I'll remove this check.